File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Table/TableItemBase.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "TableItemBase";
GlobalNS[M.clsName] = M;

function M:ctor()
    self.mItemHeader = nil;
    self.mItemBody = nil;
end

function M:parseHeaderByteBuffer(bytes)
    if(nil == self.mItemHeader) then
        self.mItemHeader = GlobalNS.new(GlobalNS.TableItemHeader);
    end
    self.mItemHeader:parseHeaderByteBuffer(bytes);
end

function M:parseBodyByteBuffer(bytes, offset, cls)
    if(nil == self.mItemBody) then
        self.mItemBody = GlobalNS.new(cls);
    end

    self.mItemBody:parseBodyByteBuffer(bytes, offset);
end

function M:parseAllByteBuffer(bytes, cls)
    -- 解析头
    self:parseHeaderByteBuffer(bytes);
    -- 保存下一个 Item 的头位置
    GlobalNS.UtilTable.mPrePos = bytes.position;
    -- 解析内容
    self:parseBodyByteBuffer(bytes, self.mItemHeader.mOffset, cls);
    -- 移动到下一个 Item 头位置
    bytes:setPos(GlobalNS.UtilTable.mPrePos);
end

return M;

File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/Table/TableItemBase.cpp
﻿#include "MyProject.h"
#include "TableItemBase.h"
#include "MByteBuffer.h"
#include "TableItemHeader.h"

TableItemBase::TableItemBase()
	: mItemHeader(nullptr), mItemBody(nullptr)
{

}

void TableItemBase::parseHeaderByteBuffer(MByteBuffer* bytes)
{
    if (nullptr == mItemHeader)
    {
		this->mItemHeader = MY_NEW TableItemHeader();
    }

	this->mItemHeader->parseHeaderByteBuffer(bytes);
}